### PR TITLE
AST: Add location to single entry rule head ref

### PR DESCRIPTION
**Pull request by Ronnie-personal**
### Why the changes in this PR are needed?
This PR is to address issue  #6199

### What are the changes in this PR?
Add location to single entry rule head ref.

### CODE REVIEW

**Comment by Ronnie-personal** [comment]
This is a draft PR, I'm still working on test cases.

**Comment by ashutosh-narkar** [comment]
> This is a draft PR, I'm still working on test cases.

Sounds good. I've converted it to a draft PR.

**Review by johanfylling — COMMENTED** [review]
Looks good 👍 , just one suggestion.

File: ast/parser_ext.go [hunk 1]
```diff
@@ -245,6 +245,9 @@ func ParseCompleteDocRuleFromEqExpr(module *Module, lhs, rhs *Term) (*Rule, erro
 
 	if v, ok := lhs.Value.(Var); ok {
 		head = NewHead(v)
+		// modify the code to add location to head ref, also need to set jsonOptions to include option
+		head.Reference[0].SetLocation(lhs.Location)
+		head.Reference[0].setJSONOptions(lhs.jsonOptions)
```
[the remark this answers is not in the document]
Comment (Ronnie-personal) [reply ast/parser_ext.go:250]: NewHead() function assigns name only to reference, so the location is missing.

`	head := &Head{
		Name:      name,
Reference: []*Term{NewTerm(name)},
	}`

In the multi-ref case, ref is directly assigned to head's Reference, so the location is carried over.
`head = RefHead(r)
...
head.Reference = r`

File: ast/parser.go [hunk 2]
```diff
@@ -864,7 +864,12 @@ func (p *Parser) parseHead(defaultRule bool) (*Head, bool) {
 
 	switch x := ref.Value.(type) {
 	case Var:
+		// modify the code to add location to head ref, also need to set jsonOptions to include option
 		head = NewHead(x)
+		head.Reference[0].SetLocation(ref.Location)
```
[the remark this answers is not in the document]
Comment (Ronnie-personal) [reply ast/parser.go:869]: Sounds good, I will update the code accordingly.

File: ast/parser.go [hunk 3]
```diff
@@ -864,7 +864,12 @@ func (p *Parser) parseHead(defaultRule bool) (*Head, bool) {
...
+		if (p.po.JSONOptions) != nil {
+			head.Reference[0].setJSONOptions(*p.po.JSONOptions)
```
[the remark this answers is not in the document]
Comment (Ronnie-personal) [reply ast/parser.go:871]: > There is [code](https://github.com/open-policy-agent/opa/blob/4164be59b9babac55a326508d8a009c179e2490b/ast/parser.go#L400) for applying json options to all parsed statements. We need to do this here despite that?
> 
When I remove the setJSONOptions, and run make test, the test failed.
--- FAIL: TestParseRulesBlockJSONOutputWithLocations (0.01s)
    parse_test.go:807: Expected output
    
    The location is missing in the ref head:
                  "head": {
                "name": "allow",
                "value": {
                  "location": {
                    "file": "/tmp/opa_test3751332167/x.rego",
                    "row": 3,
                    "col": 19
                  },
                  "type": "boolean",
                  "value": false
                },
                "ref": [
                  {		  
                    "type": "var",
                    "value": "allow"
                  }
                ],
                "location": {
                  "file": "/tmp/opa_test3751332167/x.rego",
                  "row": 3,
                  "col": 11
                }
              },

> `Parser.Parse()` seems to be the only public function leading here, so SDK cohesion regarding json option propagation shouldn't be a concern (?).

Good question! I'm not familiar with SDK cohesion. What are the existing SDK related test cases? 
I guess we could understand more about the impact if we can come up with more SDK related tests.

[the remark this answers is not in the document]
Comment (johanfylling) [reply ast/parser.go:871]: The portions of the SDK that relates to parsing are the public functions in `ast/parser.go` and `ast/parser_ext.go`, and it unfortunately doesn't look like we have a lot of tests for these. Some are however heavily used in testing other parts of OPA, as they are a convenient way to construct ASTs for "mocking". 
When I say cohesion, I'm referring to the desire to make all similar functions behave similarly; e.g. always propagating json properties, like you do here. There are no explicit tests for this that I'm aware of.

It surprises me that we start seeing failures when you remove this piece of code. Could it be that the visitor used [here](https://github.com/open-policy-agent/opa/blob/4164be59b9babac55a326508d8a009c179e2490b/ast/parser.go#L402) is neglecting some fields in the rule header 🤔 ? If that guess holds water, however, making a change there could have unintentional side effects, as this is heavily used elsewhere; so I think we can keep your changes here 👍 .

[the remark this answers is not in the document]
Comment (Ronnie-personal) [reply ast/parser.go:871]: Thanks for the detailed explanation, I will run a debug and confirm where the json options are being dropped.

[the remark this answers is not in the document]
Comment (Ronnie-personal) [reply ast/parser.go:871]: I have done the debugging, here is the findings.
This code [here](https://github.com/open-policy-agent/opa/blob/4164be59b9babac55a326508d8a009c179e2490b/ast/parser.go#L402) actually sets the jsonOptions for the stmt or the rule, it does not set the jsonOptions for the rule head ref.
In order to add the location in the 'ref' field, we need to the set jsonOptions for rule.head.reference[0].jsonOptions.

**Comment by johanfylling** [comment]
Huh, I don't know how that happened 🤔 . I definitely didn't mean to close this.

**Review by johanfylling — APPROVED** [review]
Looks good 👍 , Thank you @Ronnie-personal 😃 !

**Pull request merged**